packages/libs/deer/src/lib.rs

```diff
@@ -1,9 +1,451 @@
+    }
+    fn visit_u128(self, v: u128) -> Result<Self::Value, Self::Error> {
+        Err(Report::new(Self::Error::message(
+            "unexpected value of type u128",
+        )))
+    }
+    fn visit_usize(self, v: usize) -> Result<Self::Value, Self::Error> {
+        Err(Report::new(Self::Error::message(
+            "unexpected value of type usize",
+        )))
+    }
+
+    fn visit_f32(self, v: f32) -> Result<Self::Value, Self::Error> {
+        Err(Report::new(Self::Error::message(
+            "unexpected value of type f32",
+        )))
+    }
+    fn visit_f64(self, v: f64) -> Result<Self::Value, Self::Error> {
+        Err(Report::new(Self::Error::message(
+            "unexpected value of type f64",
+        )))
+    }
+}
+
+// internal visitor, which is used during the default implementation of the `deserialize_i*` and
+// `deserialize_u*` methods.
+struct NumberVisitor<E: Error>(PhantomData<fn() -> E>);
+
+impl<E: Error> NumberVisitor<E> {
+    fn new() -> Self {
+        Self(PhantomData::default())
+    }
+}
+
+impl<E: Error> Visitor<'_> for NumberVisitor<E> {
+    type Error = E;
+    type Value = Number;
+
+    fn expecting_display(&self) -> String {
+        "number".to_string()
+    }
+
+    fn visit_number(self, v: Number) -> Result<Self::Value, Self::Error> {
+        Ok(v)
+    }
+}
+
+macro_rules! derive_from_number {
+    [$($method:ident ($to:ident) -> $visit:ident,)*] => {
+        $(derive_from_number!(#internal, $method; $to, $visit);)*
+    };
+
+    (#internal, $method:ident; $to:ident, $visit:ident) => {
+        /// Automatically implemented convenience method, which uses [`Self::deserialize_number`]
+        /// to extract a value of the primitive type, will otherwise error out.
+        ///
+        /// # Errors
+        ///
+        /// Current value is either not a number or wasn't able to be casted to the primitive type
+        fn $method<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+        where
+            V: Visitor<'de>,
+        {
+            let n = self.deserialize_number(NumberVisitor::<Self::Error>::new())?;
+            let v = n
+                .$to()
+                .ok_or_else(|| Self::Error::message("provided value too large or too small"))
+                .into_report()?;
+
+            visitor.$visit(v).change_context(Self::Error::new())
+        }
+    };
+}
+
+/// A **data format** that can deserialize any data structure which is supported by deer.
+///
+/// This traits defines the deserialization half, while [`Deserialize`] is used to use the
+/// deserializer to convert that data into the instance of a type, or fail if that was not possible.
+///
+/// `deer`s focus is on simplicity, therefore the data model is a lot less specific than ones used
+/// by other crates (like [`serde`]).
+///
+/// The data model consists of the following types:
+///
+/// * 6 primitives:
+///     * `null`
+///         * encodes the explicit absence of a value (`undefined`/`null` in JSON)
```

Comment:
I would say that `undefined` is an absence of a field, and _generally_ should be represented by the actual field being omitted from the object (although deserializers should probably allow this to be configurable for looser interpretations).

I wouldn't explicitly say that `null` encodes both, I think that's misleading.

Reply:
This is an oversight in my initial implementation and something I explicitly set as a goal when starting 🤦, so thank you and @TimDiekmann earlier for pointing it out!

* `visit_null`: explicit absence of a value (`JS: null`)
* `visit_none`: absence of a value (`JS: undefined`)

I seem to have been quite tired while writing up the documentation to make such a trivial mistake 😅

This also enables the creation of a new type and an easy Visitor implementation for default values.

I would envision the type to look somewhat like this:

```
struct Maybe<T> {
	Some(T),
	None {
		explicit: bool
	}
}
```

(naming is TBD)

This would enable PATCH APIs to effectively model the absence of a field vs. "unsetting" of a field!

Reply:
`visit_none` added in [5f0a78a](https://github.com/hashintel/hash/pull/1170/commits/5f0a78a16c1347a59d4431a4f1d548605d99568f)